src/utils/WeakMap.js

```diff
@@ -0,0 +1,3 @@
+import WeakMap from 'es6-weak-map'
```

Comment:
I used `es6-weak-map` to polyfill `WeakMap`. If it's undesirable to add this dependency, I guess we could manually implement a fallback to using a list here - I didn't try to dive deep into the `react-proxy` discussions re polyfills/fallbacks, but `es6-weak-map` appears to have O(1) lookup time (rather than O(n) for a list)

Reply:
Yes but it [seems](https://github.com/medikoo/es6-weak-map/blob/v2.0.2/polyfill.js#L61) that this polyfill is setting a property on the weakly referenced object (in our case, the enhanced React component) so it would defeat the purpose of this PR.

I'm afraid we'll have to stick to using a list unfortunately.

If we wanted to have O(1)-ish lookup time, we could bucketize this list by taking advantage of the fact that the weakly referenced objects are components and that they (may) have a `displayName` or `name` property that we could use as bucket names.

This somehow feels a bit "clumsy" to me though. I am starting to wonder if the better way would not simply be to disable the compactable behavior of components in environments where `WeakMap` is not supported (that is, IE < 11).

How do you feel about this?

Reply:
Pinging @neoziro as he may have an opinion on this.

Reply:
@matthieuprat ah ok I thought that perhaps it was safe (hidden from `hoist-non-react-statics`) because the [referenced implementation](https://github.com/drses/weak-map) describes it as
> In a nutshell, the WeakMap shim emulates a WeakMap by adding a hidden property to the key that associates the weak map with the retained object. The shim overrides the ECMAScript 5 methods to cover its tracks.

But indeed, if I change the import from `es6-weak-map` to `es6-weak-map/polyfill` (which always uses the polyfill), tests fail

So for now removed `es6-weak-map` and implemented your suggestion of disabling compactable behavior (by making our `utils/WeakMap` not track anything if `WeakMap` does not exist) - not sure how to test the fallback behavior, if I change `export default WeakMap || NonTrackingDummyWeakMap` to `export default NonTrackingDummyWeakMap`, it is just "should be merged" tests (and `debug` test) that fail

I'm not familiar enough with `recompact` internals to know if disabling compactable behavior when `WeakMap` is missing is reasonable, but seems like it would make sense if that means that it will behave more like `recompose` (separate HOCs in the component hierarchy for each `recompact` HOC used)?

Reply:
I think it is reasonable. It will indeed behave more like recompose.

Reply:
One way of testing this behavior would be to create another test file and override the `WeakMap` global before `require`ing HOCs that directly or indirectly depend on `src/utils/WeakMap`:

```js
delete global.WeakMap
const { withProps } = require('..')

test('NonTrackingDummyWeakMap fallback', () => { /* ... */ })
```